# Data Merging and Cleaning

### Getting Started

Importing needed libraries and reading already gathered csv data.

In [1]:
import pandas as pd
from config import tmdb_api_key
import requests

In [2]:
numbers_budget_df = pd.read_csv('Data/tn.movie_budgets.csv')
tmdb_df = pd.read_csv('Data/tmdb_df.csv')

In [3]:
print(numbers_budget_df.shape)
print(tmdb_df.shape)

(5782, 6)
(4775, 14)


In [4]:
print(numbers_budget_df.columns)
print(tmdb_df.columns)

Index(['id', 'release_date', 'movie', 'production_budget', 'domestic_gross',
       'worldwide_gross'],
      dtype='object')
Index(['popularity', 'vote_count', 'video', 'poster_path', 'id', 'adult',
       'backdrop_path', 'original_language', 'original_title', 'genre_ids',
       'title', 'vote_average', 'overview', 'release_date'],
      dtype='object')


### Data Merging

We will merge on the title and year values simultaneously, as multiple movies have the same titles. We need to create these indicies to be merged on.

In [5]:
numbers_budget_df.rename(columns={'movie':'title'}, inplace=True)
numbers_budget_df['year'] = numbers_budget_df['release_date'].map(lambda x: str(x)[-4:])
tmdb_df['year'] = tmdb_df['release_date'].map(lambda x: str(x)[:4])

In [6]:
merged_df = numbers_budget_df.merge(tmdb_df, on=['title', 'year'], how='inner')
merged_df

id_x release_date_x                                        title  \
0        1   Dec 18, 2009                                       Avatar   
1        2   May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2        3    Jun 7, 2019                                 Dark Phoenix   
3        4    May 1, 2015                      Avengers: Age of Ultron   
4        7   Apr 27, 2018                       Avengers: Infinity War   
...    ...            ...                                          ...   
4359    65   Dec 31, 2007                                  Tin Can Man   
4360    67   Apr 28, 2006                                        Clean   
4361    71   Aug 14, 2008              The Rise and Fall of Miss Thang   
4362    75    Oct 8, 2004                                       Primer   
4363    79    Apr 2, 1999                                    Following   

     production_budget domestic_gross worldwide_gross  year  popularity  \
0         $425,000,000   $760,507,625  $2,776,345,279  2009      82.100   
1         $410,600,000   $241,063,875  $1,045,663,875  2011      44.410   
2         $350,000,000    $42,762,350    $149,762,350  2019     115.004   
3         $330,600,000   $459,005,868  $1,403,013,963  2015      79.895   
4         $300,000,000   $678,815,482  $2,048,134,200  2018     204.940   
...                ...            ...             ...   ...         ...   
4359           $12,000             $0              $0  2007       1.547   
4360           $10,000       $138,711        $138,711  2006       0.600   
4361           $10,000           $401            $401  2008       0.600   
4362            $7,000       $424,760        $841,926  2004      50.245   
4363            $6,000        $48,482        $240,495  1999       8.631   

      vote_count  video                       poster_path    id_y  adult  \
0          22200  False  /6EiRUJpuoeQPghrs3YNktfnqOVh.jpg   19995  False   
1          10319  False  /9zxsSzXx2UO3yNoS4j97vs4uOhc.jpg    1865  False   
2           4000  False  /cCTJPelKGLhALq3r51A9uMonxKj.jpg  320288  False   
3          16666  False  /4ssDuvEDkSArWEdyBl2X5EHvYKU.jpg   99861  False   
4          19949  False  /7WsyChQLEftFiDOVTGkv3hFpyyt.jpg  299536  False   
...          ...    ...                               ...     ...    ...   
4359           3  False  /2YMMRWzq9SzZEYau1iAtdTPN269.jpg  157185  False   
4360           0  False  /kOk47vD8E0dScAyNVNCXhsncaE4.jpg  586483  False   
4361           0  False  /2YkIO5qjfhyICn5wXhLYMJLk152.jpg  579116  False   
4362        1382  False  /xEoq2WmDzpzxhkHEsmOYOg6BPg6.jpg   14337  False   
4363         935  False  /lK9LE5D67HNyPlhJcjKAgnRh0OO.jpg   11660  False   

                         backdrop_path original_language  \
0     /bIL7ENqh1egWTxN41sM2W42JqPc.jpg                en   
1     /bfTKq6mGHt2CzZlBU31JHF9bMog.jpg                en   
2     /phxiKFDvPeQj4AbkvJLmuZEieDU.jpg                en   
3     /8i6ZDk97Vyh0jHohMG4vFeFuKJh.jpg                en   
4     /lmZFxXgJE3vgrciwuDib0N8CfQo.jpg                en   
...                                ...               ...   
4359                               NaN                en   
4360                               NaN                en   
4361                               NaN                en   
4362  /gIFbfh9O4HBrO6numxVBxtZ9fTL.jpg                en   
4363  /gCpl06aiAGUrPrmspeqhdKVQRT1.jpg                en   

                                   original_title          genre_ids  \
0                                          Avatar  [28, 12, 14, 878]   
1     Pirates of the Caribbean: On Stranger Tides       [28, 12, 14]   
2                                    Dark Phoenix      [28, 12, 878]   
3                         Avengers: Age of Ultron      [28, 12, 878]   
4                          Avengers: Infinity War      [28, 12, 878]   
...                                           ...                ...   
4359                                  Tin Can Man               [27]   
4360  

In order to analyze genre data, we will break out the genres into a column for each genre, and assign each value a boolean. TMDB has an api function to get all the genre name-to-id relationships, which we will convert into a Python dictionary.

In [7]:
tmdb_genres = requests.get(f'https://api.themoviedb.org/3/genre/movie/list?api_key={tmdb_api_key}&language=en-US').json()['genres']
tmdb_genres

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 80, 'name': 'Crime'},
 {'id': 99, 'name': 'Documentary'},
 {'id': 18, 'name': 'Drama'},
 {'id': 10751, 'name': 'Family'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 36, 'name': 'History'},
 {'id': 27, 'name': 'Horror'},
 {'id': 10402, 'name': 'Music'},
 {'id': 9648, 'name': 'Mystery'},
 {'id': 10749, 'name': 'Romance'},
 {'id': 878, 'name': 'Science Fiction'},
 {'id': 10770, 'name': 'TV Movie'},
 {'id': 53, 'name': 'Thriller'},
 {'id': 10752, 'name': 'War'},
 {'id': 37, 'name': 'Western'}]

In [8]:
genres = {}
for genre in tmdb_genres:
    genres[str(genre['id'])] = genre['name']
genres

{'28': 'Action',
 '12': 'Adventure',
 '16': 'Animation',
 '35': 'Comedy',
 '80': 'Crime',
 '99': 'Documentary',
 '18': 'Drama',
 '10751': 'Family',
 '14': 'Fantasy',
 '36': 'History',
 '27': 'Horror',
 '10402': 'Music',
 '9648': 'Mystery',
 '10749': 'Romance',
 '878': 'Science Fiction',
 '10770': 'TV Movie',
 '53': 'Thriller',
 '10752': 'War',
 '37': 'Western'}

Each genre will be looped over, creating a new column in the dataframe. Each column is mapped from the 'genre_ids' column.

In [9]:
merged_df.genre_ids

0       [28, 12, 14, 878]
1            [28, 12, 14]
2           [28, 12, 878]
3           [28, 12, 878]
4           [28, 12, 878]
              ...        
4359                 [27]
4360                   []
4361                   []
4362        [18, 878, 53]
4363             [18, 53]
Name: genre_ids, Length: 4364, dtype: object

In [10]:
for key, value in genres.items():
    merged_df[value.lower().replace(' ', '_')] = merged_df['genre_ids'].map(lambda x: key in x)

In [11]:
print(merged_df.shape)
print(merged_df.columns)

(4364, 39)
Index(['id_x', 'release_date_x', 'title', 'production_budget',
       'domestic_gross', 'worldwide_gross', 'year', 'popularity', 'vote_count',
       'video', 'poster_path', 'id_y', 'adult', 'backdrop_path',
       'original_language', 'original_title', 'genre_ids', 'vote_average',
       'overview', 'release_date_y', 'action', 'adventure', 'animation',
       'comedy', 'crime', 'documentary', 'drama', 'family', 'fantasy',
       'history', 'horror', 'music', 'mystery', 'romance', 'science_fiction',
       'tv_movie', 'thriller', 'war', 'western'],
      dtype='object')


### Data Cleaning

Movies that have no genre or revene data will not be useful in our analysis, so we will remove them. Dupliactes will also be removed. This only affects 205 data points.

In [12]:
no_genres = list(merged_df[merged_df['genre_ids'] == '[]'].index)
len(no_genres)

20

In [13]:
merged_df.drop(index=no_genres, inplace=True)

In [14]:
no_gross = list(merged_df[merged_df['worldwide_gross'] == '$0'].index)
len(no_gross)

180

In [15]:
merged_df.drop(index=no_gross, inplace=True)

In [16]:
duplicates = list(merged_df[merged_df.duplicated()].index)
len(duplicates)

8

In [17]:
merged_df.drop(index=duplicates, inplace=True)

In [18]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4156 entries, 0 to 4363
Data columns (total 39 columns):
id_x                 4156 non-null int64
release_date_x       4156 non-null object
title                4156 non-null object
production_budget    4156 non-null object
domestic_gross       4156 non-null object
worldwide_gross      4156 non-null object
year                 4156 non-null object
popularity           4156 non-null float64
vote_count           4156 non-null int64
video                4156 non-null bool
poster_path          4140 non-null object
id_y                 4156 non-null int64
adult                4156 non-null bool
backdrop_path        3974 non-null object
original_language    4156 non-null object
original_title       4156 non-null object
genre_ids            4156 non-null object
vote_average         4156 non-null float64
overview             4156 non-null object
release_date_y       4156 non-null object
action               4156 non-null bool
adventure         

Many of the columns are unecessary, and some have unhelpful data types. We will also create a few new columns from the existing that may be useful, and rename ones with unhelpful titles.

In [19]:
merged_df.drop(columns = ['id_x', 'release_date_x', 'id_y', 'poster_path', 'adult', 'backdrop_path', 'video'], inplace=True)

In [20]:
merged_df['production_budget'] = merged_df['production_budget'].map(lambda x: int(str(x).strip('$').replace(',', '')))
merged_df['domestic_gross'] = merged_df['domestic_gross'].map(lambda x: int(str(x).strip('$').replace(',', '')))
merged_df['worldwide_gross'] = merged_df['worldwide_gross'].map(lambda x: int(str(x).strip('$').replace(',', '')))

In [21]:
merged_df['worldwide_profit(loss)'] = merged_df['worldwide_gross'] - merged_df['production_budget']

In [22]:
merged_df['roi'] = merged_df['worldwide_profit(loss)'] * 100 / merged_df['production_budget']

In [23]:
merged_df['year'] = merged_df['year'].map(int)
merged_df['month'] = merged_df['release_date_y'].map(lambda x: int(x.split('-')[1]))
merged_df['day'] = merged_df['release_date_y'].map(lambda x: int(x.split('-')[2]))

In [24]:
merged_df.rename(columns = {'release_date_y': 'release_date'}, inplace = True)

In [25]:
merged_df.columns

Index(['title', 'production_budget', 'domestic_gross', 'worldwide_gross',
       'year', 'popularity', 'vote_count', 'original_language',
       'original_title', 'genre_ids', 'vote_average', 'overview',
       'release_date', 'action', 'adventure', 'animation', 'comedy', 'crime',
       'documentary', 'drama', 'family', 'fantasy', 'history', 'horror',
       'music', 'mystery', 'romance', 'science_fiction', 'tv_movie',
       'thriller', 'war', 'western', 'worldwide_profit(loss)', 'roi', 'month',
       'day'],
      dtype='object')

Rearanging the columns makes the data easier to read.

In [26]:
cleaned_df = merged_df[['title','original_title', 'year', 'month', 'day', 'release_date', 'production_budget', 'domestic_gross', 'worldwide_gross', 
                           'worldwide_profit(loss)', 'roi', 'popularity', 'vote_count', 'vote_average', 'original_language', 'genre_ids',
        'action', 'adventure', 'animation', 'comedy', 'crime',
       'documentary', 'drama', 'family', 'fantasy', 'history', 'horror',
       'music', 'mystery', 'romance', 'science_fiction', 'tv_movie',
       'thriller', 'war', 'western', 'overview']]

In [27]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4156 entries, 0 to 4363
Data columns (total 36 columns):
title                     4156 non-null object
original_title            4156 non-null object
year                      4156 non-null int64
month                     4156 non-null int64
day                       4156 non-null int64
release_date              4156 non-null object
production_budget         4156 non-null int64
domestic_gross            4156 non-null int64
worldwide_gross           4156 non-null int64
worldwide_profit(loss)    4156 non-null int64
roi                       4156 non-null float64
popularity                4156 non-null float64
vote_count                4156 non-null int64
vote_average              4156 non-null float64
original_language         4156 non-null object
genre_ids                 4156 non-null object
action                    4156 non-null bool
adventure                 4156 non-null bool
animation                 4156 non-null bool
comedy     

In [28]:
cleaned_df

title  \
0                                          Avatar   
1     Pirates of the Caribbean: On Stranger Tides   
2                                    Dark Phoenix   
3                         Avengers: Age of Ultron   
4                          Avengers: Infinity War   
...                                           ...   
4351                        In the Company of Men   
4352                                       Exeter   
4354                                      Slacker   
4362                                       Primer   
4363                                    Following   

                                   original_title  year  month  day  \
0                                          Avatar  2009     12   10   
1     Pirates of the Caribbean: On Stranger Tides  2011      5   14   
2                                    Dark Phoenix  2019      6    5   
3                         Avengers: Age of Ultron  2015      4   22   
4                          Avengers: Infinity War  2018      4   25   
...                                           ...   ...    ...  ...   
4351                        In the Company of Men  1997      3   28   
4352                                       Exeter  2015      1   16   
4354                                      Slacker  1991      7    5   
4362                                       Primer  2004     10    8   
4363                                    Following  1999      4    2   

     release_date  production_budget  domestic_gross  worldwide_gross  \
0      2009-12-10          425000000       760507625       2776345279   
1      2011-05-14          410600000       241063875       1045663875   
2      2019-06-05          350000000        42762350        149762350   
3      2015-04-22          330600000       459005868       1403013963   
4      2018-04-25          300000000       678815482       2048134200   
...           ...                ...             ...              ...   
4351   1997-03-28              25000         2883661          2883661   
4352   2015-01-16              25000               0           489792   
4354   1991-07-05              23000         1227508          1227508   
4362   2004-10-08               7000          424760           841926   
4363   1999-04-02               6000           48482           240495   

      worldwide_profit(loss)  ...  horror  music  mystery  romance  \
0                 2351345279  ...   False  False    False    False   
1                  635063875  ...   False  False    False    False   
2                 -200237650  ...   False  False    False    False   
3                 1072413963  ...   False  False    False    False   
4                 1748134200  ...   False  False    False    False   
...                      ...  ...     ...    ...      ...      ...   
4351                 2858661  ...   False  False    False    False   
4352                  464792  ...    True  False    False    False   
4354                 1204508  ...   False  False    False    False   
4362                  834926  ...   False  False    False    False   
4363                  234495  ...   False  False    False    False   

     science_fiction tv_movie  thriller    war  western  \
0               True    False     False  False    False   
1              False    False     False  False    False   
2               True    False     False  False    False   
3               True    False     False  False    False   
4               True    False     False  False    False   
...              ...      ...       ...    ...      ...   
4351           False    False     False  False    False   
4352           False    False      True  False    False   
4354           False    False     False  False    False   
4362            True    False      True  False    False   
4363           False    False      True  False    False   

                                               overview  
0     In the 22nd century, a paraplegic Marine is di...  
1     Capta

Saving the cleaned and merged dataframe to a csv file.

In [29]:
cleaned_df.to_csv('Data/cleaned_data.csv', index=False)